python/phase-1.md

```diff
@@ -0,0 +1,460 @@
+# Python SDK - Phase 1
+
+The Python SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 1. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+The Python SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing a bridge based on
+[PyO3](https://github.com/PyO3/PyO3) similar to `sdk-typescript`'s `neon`-based bridge.
+
+The high-level goals for the Python SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as reasonable with regards to dependencies
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+
+async def main():
+    async with temporalio.Client("localhost:7233") as client:
+        result = await client.execute_workflow(
+            "MyWorkflow", "arg1", id="someid", task_queue="somequeue"
+        )
+        print("Result", result)
+```
+
+An example of an activity and registration (also missing a lot of features):
+
+```python
+import temporalio
+
+
+def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+
+async def main():
+    async with temporalio.Client("localhost:7233") as client:
+        worker = temporalio.Worker(
+            client, task_queue="somequeue", activities={"say_hello": say_hello}
+        )
+        await worker.run()
+```
+
+## API Definitions
+
+A lot of loosely-defined API definitions are below. They are intentionally lacking details.
+
+High-level notes:
+
+* For demonstration/simplicity here we are leaving off a lot of proper typings (e.g. `Optional`) that the real version
+  will have
+* We leave off a lot of options on calls for this high-level proposal
+* We are not using proper function documentation, but rather inline comments for easier discussion
+* **NOTE: These will have proper docs and types when developed!**
+
+### Client
+
+Exported from `temporalio` (but probably in `temporalio/client/client.py`, exported from `temporalio.client` then
```

Comment:
I am thinking this may not be clearest exported from `temporalio` directly, but instead leave only in the `client` module. So you would have to `from temporalio import client` and use `client.Client`. This more aligns with our expectations of using `from temporalio import activity` for e.g. `activity.info()`. Thought I do think all common code such as retries and errors may deserve to be in the `temporalio` package directly.

Thoughts?